engine/schema/src/main/java/com/cloud/upgrade/dao/DatabaseAccessObject.java

```diff
@@ -26,6 +26,17 @@ public class DatabaseAccessObject {
 
     private static Logger s_logger = Logger.getLogger(DatabaseAccessObject.class);
 
+    public void addForeignKey(Connection conn, String tableName, String tableColumn, String foreignTableName, String foreignColumnName) {
+        String addForeignKeyStmt = String.format("CALL `cloud`.`IDEMPOTENT_ADD_FOREIGN_KEY`('%s', '%s', '%s', '%s')", tableName, tableColumn, foreignTableName, foreignColumnName);
```

Comment:
@DaanHoogland 
my suggestion is (do not commit it)
```suggestion
        String addForeignKeyStmt = String.format("ALTER TABLE xxx ADD CONSTRAINT xxxxx xxxxx");
```
with the changes in schema-41720to41800.sql, the procedure `IDEMPOTENT_ADD_FOREIGN_KEY` is not used at all, so we can remove it to avoid it being used in the future

cc @harikrishna-patnala

Reply:
ping @harikrishna-patnala will you be addressing this ^^ cc @rohityadavcloud

Reply:
I am not sure it is wise to allow changes to the DB to be not IdemPotent if they can be. updating in dot releases that rquire updates will lead to conflicts if we do not keep this policy. I have not understood yet why it would be harmful in this instance. @weizhouapache please explain.

Reply:
> I am not sure it is wise to allow changes to the DB to be not IdemPotent if they can be. updating in dot releases that rquire updates will lead to conflicts if we do not keep this policy. I have not understood yet why it would be harmful in this instance. @weizhouapache please explain.

@DaanHoogland 
the try-catch block will handle the exceptions if the column or constrain key has been already added.

I think it is better than MySQL procedure, because the error codes of MySQL procedure are different with MySQL versions, which is the root cause of #7358 

```
    try
        {
            pstmt.executeUpdate();
            s_logger.debug(String.format("Foreign key is added successfully from the table %s", tableName));
        } catch (SQLException e) {
            s_logger.error("Ignored SQL Exception when trying to add foreign key on table "  + tableName + " exception: " + e.getMessage());
        }
    }
}

Reply:
Agree with @weizhouapache it is happening because of the different error codes. 

Sorry for the delay @shwstppr. I'm on this now.